### ISSUE TITLE
Service as separated K8S Pod/Deployment

### ISSUE THREAD

**Issue by echarles** [jupyterhub/zero-to-jupyterhub-k8s#2154]
Elaboraring a bit more on https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2148 https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2132, I was wondering if anyone feels the need to have a way to launch jupyterhub services as separated pod (or Deployment)?

I am seeing a case where the hub is the central place for Auth, and the services satellites providing features to the user.

**Comment by consideRatio** [comment]
I consider it very relevant and useful to launch JupyterHub registered services running external to the jupyterhub pod as well!

Technically I think there is support for doing this already in the Helm chart, and I think of the PRs you mentioned was a way to complement this support with JupyterHub registered services running in the same container as JupyterHub.

I consider it out of scope for this Helm chart to generate general k8s Deployment resources for JupyterHub registered external services, but... maybe it could be in scope to let this Helm chart support creating a layer for authentication/authorization and network access for users to some JupyterHub un-aware service in a JupyterHub.

I'm thinking for example about a web server hosting documentation about the JupyterHub instance. If such web server was hosted and not accessible publicly but internally in the k8s cluster, I'm entertaining the idea if this Helm chart for example could assist wrapping that web server to be publicly accessible to JupyterHub users. This would mean JupyterHub based authentication, some authorization logic somewhere, and networking to enforce that access.

**Comment by echarles** [comment]
> Technically I think there is support for doing this already in the Helm chart, and I think of the PRs you mentioned was a way to complement this support with JupyterHub registered services running in the same container as JupyterHub.

I have scanned the specs in https://zero-to-jupyterhub.readthedocs.io/en/latest/resources/reference.html and don't see a place where I could define that... @consideRatio If you could link to maybe the source to do this, that would be super helpful. 

> I consider it out of scope for this Helm chart to generate general k8s Deployment resources for JupyterHub registered external services, but... maybe it could be in scope to let this Helm chart support creating a layer for authentication/authorization and network access for users to some JupyterHub un-aware service in a JupyterHub.

I love the idea to provide the Auth layer. This would be useful for a lot of case. BTW I am working on Jupyter RTC project for which we have a workshop (on various topics, including Auth) this Monday and we'd love to have JupyterHub folks around to help with this https://hackmd.io/UbnBH58hS8itoWgfiWT77A?both#Agenda.

What do you mean with `network access for users to some JupyterHub un-aware service in a JupyterHub`? Are you referring to the routes maintained by the JupyterHub configurable HTTP proxy?

> I'm thinking for example about a web server hosting documentation about the JupyterHub instance. If such web server was hosted and not accessible publicly but internally in the k8s cluster, I'm entertaining the idea if this Helm chart for example could assist wrapping that web server to be publicly accessible to JupyterHub users. This would mean JupyterHub based authentication, some authorization logic somewhere, and networking to enforce that access.

Auth just means providing the Pods with a JupyteHub Token, right? Networking is configuring the CHP, right?

> Action points Explore ideas and such but clarify them concretely enough to be viable to evaluate for actual implementation

Who?

**Comment by minrk** [comment]
It's definitely appropriate to deploy services as separate k8s resources, and the right choice for almost any service, but especially stateful ones that serve UI. I also agree with @consideRatio that it shouldn't be in scope for this chart to deploy the services themselves. An *example* in the docs showing how to do it seems appropriate, though, and if there are hurdles in how this chart works to doing that, we can try to address those.

BinderHub (with auth enabled) is one example of a JupyterHub service deployed as a separate pod in its own chart.

**Comment by echarles** [comment]
>  I also agree with @consideRatio that it shouldn't be in scope for this chart to deploy the services themselves. 

@minrk What do you think of what @consideRatio mentioned `maybe it could be in scope to let this Helm chart support creating a layer for authentication/authorization and network access for users to some JupyterHub un-aware service in a JupyterHub.`- Would you consider exploring that road, although I don't see yet concretely how it would be implemented.

**Comment by consideRatio** [comment]
> @consideRatio If you could link to maybe the source to do this, that would be super helpful.

`hub.services` allow you to define a JupyterHub registered service and associated api-token for it to use against JupyterHub. This token would then be something you would use to speak with the JupyterHub's API. The Helm chart doesn't contain much details on that, as it's JupyterHub config just passed through. This configuration is documented at https://jupyterhub.readthedocs.io/en/stable/reference/services.html

> this Monday and we'd love to have JupyterHub folks around to help with this

I've added a calendar entry for the meet! I'll absolutely join!

> What do you mean with `network access for users to some JupyterHub un-aware service in a JupyterHub`? Are you referring to the routes maintained by the JupyterHub configurable HTTP proxy?

If you have a JupyterHub external service, how is that different from any service? I'm thinking:
1. It could passively expose itself only to jupyterhub users
2. It could actively communicate with the JupyterHub API in various ways based on user interactions etc.

Now to your question. I meant that we would be able to address the first point above - for passive services. For the second point, the service needs to be JupyterHub aware and do the requests itself rather than be wrapped by a verify access checkpoint.

Having thought a bit more about this, I now think that there are already finished solutions for accomplishing the first point quite well, so to embed it into z2jh seems to add too much complexity with too little benefit I think at this point. For example [having an k8s Ingress controller (nginx) that inspects the Ingress resources and based on annotations also ensures access is authorized against some OAuth2 identity provider](https://kubernetes.github.io/ingress-nginx/examples/auth/oauth-external-auth/#example-oauth2-proxy-kubernetes-dashboard).

A related JupyterHub example is available [here](https://github.com/jupyterhub/jupyterhub/blob/master/examples/external-oauth/README.md).

> Auth just means providing the Pods with a JupyteHub Token, right? Networking is configuring the CHP, right?

To summarize my considered example with some parts:

1. JupyterHub
   We register a JupyterHub service so we can use an API token to communicate with JupyterHub about what users are available etc.
2. Auth proxy
   We deploy software that proxies traffic to "some web server" only if the user successfully is authenticated against JupyterHub
3. Some web server
   This web server can do whatever, but it won't be aware of the user arriving has authenticated already. So, if this web server wants to make use of a user's identity, I don't have a clear idea on doing that other than doing it all the way from scratch which pretty much makes the auth proxy a redundant step.

**Comment by echarles** [comment]
> hub.services allow you to define a JupyterHub registered service and associated api-token for it to use against JupyterHub.

But those services are just added in the hub config, so the processes remain local (inside) to the hub pod. They are not used to create separated Pods/Deployments hosting those services.

I am thinking more to service definition being a docker image and a number of replicas.

> I've added a calendar entry for the meet! I'll absolutely join!

👍 

> If you have a JupyterHub external service, how is that different from any service? I'm thinking:
> It could passively expose itself only to jupyterhub users
> It could actively communicate with the JupyterHub API in various ways based on user interactions etc.

I see those options a variants... actually a service could have both: expose to users + communicate with the hub api.

Then I am a bit lost with your link to the nginx ingress resource - Let me reread that tomorrow...

> This web server can do whatever, but it won't be aware of the user arriving has authenticated already. So, if this web server wants to make use of a user's identity, I don't have a clear idea on doing that other than doing it all the way from scratch which pretty much makes the auth proxy a redundant step.

I believe this is part of the goal of Monday workshop. Let's catch up there... if no one has more to say here before...

Thx a lot for the inputs and brainstorming so far!

**Comment by consideRatio** [comment]
> But those services are just added in the hub config, so the processes remain local (inside) to the hub pod. They are not used to create separated Pods/Deployments hosting those services.

There is no process started in the hub pod at all unless you specify `command` under your `hub.service` entry.

> I am thinking more to service definition being a docker image and a number of replicas.

I think it is out of scope for any Helm chart to assist its users to deploy standalone k8s resources unrelated to the chart itself if the only integration aspect was passing a token or similar. A `hub.services` registered token can already be accessed by external JupyterHub services deployed in k8s by mounting the a JupyterHub Helm chart managed k8s Secret - it contains the tokens exposed within it.

> I see those options a variants... actually a service could have both: expose to users + communicate with the hub api.

We agree! I categorized them as such as think only a fully passive service can get sufficient help from an auth proxy, while any kind of active service would need to do the auth proxy equivalent of work anyhow.

Looking forward to the workshop on Monday! I'm watching old meeting videos now to be better up to date with deliberations and the RTC work in general.

**Comment by echarles** [comment]
> I think it is out of scope for any Helm chart to assist its users to deploy standalone k8s resources unrelated to the chart itself if the only integration aspect was passing a token or similar.

Agree!

> A hub.services registered token can already be accessed by external JupyterHub services deployed in k8s by mounting the a JupyterHub Helm chart managed k8s Secret - it contains the tokens exposed within it.

I will have a closer look and try that out. Thx for the hint.

> Looking forward to the workshop on Monday! I'm watching old meeting videos now to be better up to date with deliberations and the RTC work in general.

We have not been good at recording and publishing all the materials. Anyway, so far, we have not discussed about Auth. I have put some notes in this document, feel free to update it as you feel.

https://docs.google.com/document/d/1U9TRpDzkv7IIY3l13rw5Vd3xq0d4DMxpWS5jjjzVovI/edit

**Comment by minrk** [comment]
> maybe it could be in scope to let this Helm chart support creating a layer for authentication/authorization and network access for users to some JupyterHub un-aware service in a JupyterHub.

Definitely possible. The JupyterHub auth proxy needs to exist, first. I definitely don't think it should be *defined* in this repo, but I do think it should exist! It should be a standalone package, though - basically a generalization of jupyter-server-proxy.

That said, I *expect* it will be simpler for everyone if, once that exists, services add the auth layer to their pods as a sidecar container, rather than having a third pod that only proxies communication from the hub to the service pod with auth. But we can definitely explore both options when we get to it.

Advantages of putting the auth in the service pod:

- not-hub-aware service doesn't need to accept network traffic from outside the pod, as only the auth proxy is exposed
- one less pod to manage
- service gets to configure the proxy in the way that works best for it
- As far as JupyterHub is aware, there's no difference between a two-container service and a one-container one


> I will have a closer look and try that out. Thx for the hint.

[BinderHub is an example](https://github.com/jupyterhub/binderhub/blob/e2ebb7a948bc8e120d9b1ec422ec25b048d5831a/helm-chart/binderhub/values.yaml#L127-L130). Defining external services in jupyterhub config is often only registering the api token and permissions for the service, leaving it up to kubernetes/whatever to managed the process itself. This is the difference between "[hub-managed](https://jupyterhub.readthedocs.io/en/latest/reference/services.html#hub-managed-services)" and "[externally-managed](https://jupyterhub.readthedocs.io/en/latest/reference/services.html#externally-managed-services)" services, in Hub's terminology.

**Comment by consideRatio** [comment]
@echarles agree to close this? I'm very interested to work towards helping a z2jh based JupyterHub distribution work well with RTC, and I think the key action point that I'd work on at this point would be to explore and hash out what kind of UX we expect from a RTC enabled JupyterHub deployment on a quite high level.

I opened https://github.com/jupyterlab/jupyterlab/issues/10119 as a placeholder for such discussion at this point, understanding jupyterlab/jupyterlab + the RTC github issue label would be the place to do that.

**Comment by echarles** [comment]
@consideRatio Yes, I am closing this. 

 jupyterlab/jupyterlab + the RTC github issue label is the current consigne, we may need to adapt this rule as there are very jupyterhub specific things to be discussed and implemented to make RTC happen.

BTW this issue was not really RTC focussed, it was more a general question on having some "services" in separated pods.
